### PR TITLE
[example]support ViT with ColoTensor

### PR DESCRIPTION
**Pull request by ZhaoYi1222**

### CODE REVIEW

File: image/vision_transformer/colo_vit/test_vit.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from colossalai.nn.parallel.data_parallel import ColoDDP
+
+
+def init_1d_row_spec(model):
+    spec = TensorSpec(
+        distspec.shard(gpc.get_group(ParallelMode.PARALLEL_1D), [-1], [gpc.get_world_size(ParallelMode.PARALLEL_1D)]),
+        ParallelAction(ComputePattern.TP1D))
+    with DistSpecManager.no_grad():
+        for n, p in model.named_parameters():
+            if 'weight' in n and 'layernorm' not in n and 'embeddings.patch_embeddings.projection.weight' not in n:
```
[the remark this answers is not in the document]
Comment (Wesley-Jzy) [reply image/vision_transformer/colo_vit/test_vit.py:25]: The function name is not clear though it will not make a wrong result since every output is gathered if the gather_out arg is not set. In fact, this function sets TP_1D for dim=-1(the last dim). Only for all Linear, it's 1d_row split because Linear will be transposed when calculating. But for other layers, it's 1d_col split.

File: image/vision_transformer/colo_vit/test_vit.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                p.set_spec(spec)
+
+
+def init_1d_col_spec(model):
```
Comment (Wesley-Jzy) [inline image/vision_transformer/colo_vit/test_vit.py:29]: similarly, it's col split for Linear but row split for others.

File: image/vision_transformer/colo_vit/test_vit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZhaoYi1222) [reply image/vision_transformer/colo_vit/test_vit.py:25]: The result is incorrect with 'projection.weight'.
I follow the implementation from tests/test_tensor/test_gpt.py to get test_vit.py
I am not familiar with detailed implementation of ColoTensor, but it seems that it can pass all the tests.
Any suggestions for further revision?

[the remark this answers is not in the document]
Comment (Wesley-Jzy) [reply image/vision_transformer/colo_vit/test_vit.py:25]: e.g. to find out why projection.weight doesn't work?

File: image/vision_transformer/colo_vit/test_vit.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                          process_group=gpc.get_group(ParallelMode.DATA))
+    for torch_p, p in zip(torch_model.parameters(), model.parameters()):
+        torch_p.data.copy_(p)
+    init_spec_func(model)
+
+    check_param_equal(model, torch_model)
+    model.train()
+    torch_model.train()
+    set_seed(gpc.get_local_rank(ParallelMode.DATA))
+    for i, image_dict in enumerate(train_dataloader):
```
[the remark this answers is not in the document]
Comment (ZhaoYi1222) [reply image/vision_transformer/colo_vit/test_vit.py:87]: torch.optim.Adam is added in this example.

**Review by Wesley-Jzy — APPROVED** [review]

**Review by ver217 — APPROVED** [review]
